sway-core/src/asm_generation/from_ir.rs

```diff
@@ -1639,16 +1654,70 @@ impl<'ir> AsmBuilder<'ir> {
             }
         }
     }
+}
+
+fn ir_constant_to_ast_literal(constant: &Constant) -> Literal {
+    match &constant.value {
+        ConstantValue::Undef => unreachable!("Cannot convert 'undef' to a literal."),
+        ConstantValue::Unit => Literal::U64(0), // No unit.
+        ConstantValue::Bool(b) => Literal::Boolean(*b),
+        ConstantValue::Uint(n) => Literal::U64(*n),
+        ConstantValue::B256(bs) => Literal::B256(*bs),
+        ConstantValue::String(str) => Literal::String(crate::span::Span {
+            span: pest::Span::new(std::sync::Arc::from(str.as_str()), 0, str.len()).unwrap(),
+            path: None,
+        }),
+        ConstantValue::Array(_) => unimplemented!(),
+        ConstantValue::Struct(_) => unimplemented!(),
+    }
+}
+
+// -------------------------------------------------------------------------------------------------
+
+#[derive(Default)]
+pub struct TypeAnalyzer {
+    // The layouts of each aggregate; their whole size in bytes and field offsets in words.
+    aggregate_layouts: HashMap<Aggregate, (u64, Vec<FieldLayout>)>,
+}
+
+pub struct FieldLayout {
+    pub offset_in_words: u64, // Use words because LW/SW do.
+    pub size_in_bytes: u64,   // Use bytes because CFEI/MCP do.
+}
+
+impl TypeAnalyzer {
+    pub fn ir_type_size_in_bytes(&mut self, context: &Context, ty: &Type) -> u64 {
+        match ty {
+            Type::Unit | Type::Bool | Type::Uint(_) => 8,
+            Type::B256 => 32,
+            Type::String(n) => *n,
+            Type::Array(aggregate) | Type::Struct(aggregate) => {
+                self.analyze_aggregate(context, aggregate);
+                self.aggregate_size(context, aggregate)
+            }
+            Type::Union(aggregate) => {
+                self.analyze_aggregate(context, aggregate);
+                self.aggregate_max_field_size(context, aggregate)
+            }
+            Type::ContractCaller(_) => {
+                // We only store the address.
+                32
+            }
+            Type::Contract => {
+                unimplemented!("do contract/contract caller have/need a size?")
```

Comment:
Remove `unimplemented!()`?

Reply:
`Contract` isn't a real type, it's a bit weird.  I've given it a size of 0.